src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
that would work too. it is a bit different pattern than what we have now - we only provide mandatory one as abstract and optional stuff as a method call with lambda so that user doesnt need to provide empty override just to fulfill language requirement.